File: Submariner/SBPlaylistController.m
//
//  SBPlaylistController.m
//  Submariner
//
//  Created by Rafaël Warnault on 06/06/11.
//
//  Copyright (c) 2011-2014, Rafaël Warnault
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of the Read-Write.fr nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "SBPlaylistController.h"
#import "SBDatabaseController.h"

#import "Submariner-Swift.h"


@implementation SBPlaylistController


@synthesize playlist;
@synthesize playlistSortDescriptors;
@synthesize databaseController;



#pragma mark - 
#pragma mark LifeCycle

+ (NSString *)nibName {
    return @"Playlist";
}


- (NSString*)title {
    return [NSString stringWithFormat: @"Playlist \"%@\"", self.playlist.resourceName];
}


- (id)initWithManagedObjectContext:(NSManagedObjectContext *)context {
    self = [super initWithManagedObjectContext:context];
    if (self) {
        NSSortDescriptor *desc = [NSSortDescriptor sortDescriptorWithKey:@"playlistIndex" ascending:YES];
        playlistSortDescriptors = [NSArray arrayWithObject:desc];
    }
    return self;
}




- (void)loadView {
    [super loadView];
    
    [tracksTableView registerForDraggedTypes: @[SBLibraryTableViewDataType, SBLibraryItemTableViewDataType]];
    
    [tracksController addObserver:self
                      forKeyPath:@"selectedObjects"
                      options:NSKeyValueObservingOptionNew
                      context:nil];
}


- (void)dealloc {
    [tracksController removeObserver:self forKeyPath:@"selectedObjects"];
}


- (void)viewDidAppear {
    [super viewDidAppear];
    [[NSNotificationCenter defaultCenter] postNotificationName: @"SBTrackSelectionChanged"
                                                        object: tracksController.selectedObjects];
}


- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    if (object == tracksController && [keyPath isEqualToString:@"selectedObjects"] && self.view.window != nil) {
        [[NSNotificationCenter defaultCenter] postNotificationName: @"SBTrackSelectionChanged"
                                                            object: tracksController.selectedObjects];
    }
}




#pragma mark - 
#pragma mark Utils

- (void)clearPlaylist {
    //[tracksController setContent:nil];
}


#pragma mark - 
#pragma mark IBActions

- (IBAction)trackDoubleClick:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    if(selectedRow != -1) {
        [[SBPlayer sharedInstance] playTracks: [tracksController arrangedObjects] startingAt: selectedRow];
    }
}

- (IBAction)playSelected:(id)sender {
    [self trackDoubleClick:sender];
}

- (IBAction)addSelectedToTracklist:(id)sender {
    NSIndexSet *indexSet = [tracksTableView selectedRowIndexes];
    NSMutableArray *tracks = [NSMutableArray array];
    
    [indexSet enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
        [tracks addObject:[[tracksController arrangedObjects] objectAtIndex:idx]];
    }];
    
    [[SBPlayer sharedInstance] addTrackArray:tracks replace:NO];
}

- (IBAction)removeTrack:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    
    if(selectedRow != -1) {
        SBTrack *selectedTrack = [[tracksController arrangedObjects] objectAtIndex:selectedRow];
        if(selectedTrack != nil) {
            NSAlert *alert = [[NSAlert alloc] init];
            [alert addButtonWithTitle:@"Remove"];
            [alert addButtonWithTitle:@"Cancel"];
            [alert setMessageText:@"Remove the selected tracks?"];
            [alert setInformativeText:@"The selected tracks will be removed from this playlist."];
            [alert setAlertStyle:NSAlertStyleWarning];
            
            [alert beginSheetModalForWindow: [[self view] window] completionHandler:^(NSModalResponse returnCode) {
                [self removeTrackAlertDidEnd: alert returnCode: returnCode contextInfo: nil];
            }];
        }
    }
}


- (IBAction)delete:(id)sender {
    [self removeTrack: sender];
}


- (void)removeTrackAlertDidEnd:(NSAlert *)alert returnCode:(NSInteger)returnCode contextInfo:(void *)contextInfo {
    if (returnCode == NSAlertFirstButtonReturn) {
        NSIndexSet *selectedRows = [tracksTableView selectedRowIndexes];
        
        // delete each indiviually
        [selectedRows enumerateIndexesWithOptions: NSEnumerationReverse usingBlock:^(NSUInteger idx, BOOL * _Nonnull stop) {
            SBTrack *selectedTrack = [[tracksController arrangedObjects] objectAtIndex: idx];
            if (selectedTrack != nil) {
                [playlist removeTracksObject:selectedTrack];
            }
        }];
        
        // delete all the rows at once in the server
        if (playlist.server) {
            NSArray *selectedRowsArray = [selectedRows toArray];
            [playlist.server updatePlaylistWithID: playlist.itemId
                                             name: nil
                                          comment: nil
                                        appending: nil
                                         removing: selectedRowsArray];
        }
    }
}


- (IBAction)showSelectedInFinder:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    [self showTracksInFinder: tracksController.arrangedObjects selectedIndices: tracksTableView.selectedRowIndexes];
}


- (IBAction)downloadSelected:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    
    if(selectedRow != -1) {
        [self downloadTracks: tracksController.arrangedObjects selectedIndices: tracksTableView.selectedRowIndexes databaseController: nil];
    }
}


- (IBAction)createNewLocalPlaylistWithSelectedTracks:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    [self createLocalPlaylistWithSelected: tracksController.arrangedObjects selectedIndices: tracksTableView.selectedRowIndexes databaseController: nil];
}


- (IBAction)showSelectedInLibrary:(id)sender {
    NSInteger selectedRow = [tracksTableView selectedRow];
    
    if(selectedRow == -1) {
        return;
    }
    
    // only makes sense to have a single track, imho
    NSUInteger index = tracksTableView.selectedRowIndexes.firstIndex;
    SBTrack *track = (SBTrack*)[tracksController.arrangedObjects objectAtIndex: index];
    [[self databaseController] goToTrack: track];
}




#pragma mark -
#pragma mark NSTableView (Drag & Drop)

- (id<NSPasteboardWriting>)tableView:(NSTableView *)tableView pasteboardWriterForRow:(NSInteger)row {
    if (tableView == tracksTableView) {
        SBTrack *track = tracksController.arrangedObjects[row];
        return [[SBLibraryItemPasteboardWriter alloc] initWithItem: track index: row];
    }
    return nil;
}

- (NSDragOperation)tableView:(NSTableView*)tv validateDrop:(id <NSDraggingInfo>)info proposedRow:(NSInteger)row proposedDropOperation:(NSTableViewDropOperation)op
{
    if (row == -1)
        return NSDragOperationNone;
    
    if (op == NSTableViewDropAbove) {
        // internal drop track
        if ([info.draggingPasteboard libraryItems] != nil) {
            return NSDragOperationMove;
        }
    }
    
    return NSDragOperationNone;
}


- (BOOL)tableView:(NSTableView *)aTableView acceptDrop:(id <NSDraggingInfo>)info
              row:(NSInteger)row dropOperation:(NSTableViewDropOperation)operation
{
    NSArray<SBTrack*> *tracks = [info.draggingPasteboard libraryItemsWithManagedObjectContext: self.managedObjectContext];
    
    // internal drop track
    if (tracks != nil) {
        NSInteger sourceRow = 0;
        NSInteger destinationRow = 0;
        
        // update playlist indexes 
        if([[[tracks objectAtIndex:0] playlistIndex] integerValue] < row) {
            sourceRow = [[[tracks objectAtIndex:0] playlistIndex] integerValue];
            destinationRow = row;
            
            // increment interval rows
            NSArray *trackInInterval = [[tracksController arrangedObjects] subarrayWithRange:NSMakeRange(sourceRow, destinationRow-sourceRow)];
            for(SBTrack *track in trackInInterval) {
                NSInteger playlistIndex = [[track playlistIndex] integerValue];
                playlistIndex--;
                
                [track setPlaylistIndex:[NSNumber numberWithInteger:playlistIndex]];
            }
        } else {
            sourceRow = row;
            destinationRow = [[[tracks objectAtIndex:0] playlistIndex] integerValue];
            
            // increment interval rows
            NSArray *trackInInterval = [[tracksController arrangedObjects] subarrayWithRange:NSMakeRange(sourceRow, destinationRow-sourceRow)];
            for(SBTrack *track in trackInInterval) {
                NSInteger playlistIndex = [[track playlistIndex] integerValue];
                playlistIndex++;
                
                [track setPlaylistIndex:[NSNumber numberWithInteger:playlistIndex]];
            }
        }
    
        
        // reverse track array
        NSArray<SBTrack*> *reversedArray = [[tracks reverseObjectEnumerator] allObjects];
        
        // add reversed track at index
        for(SBTrack *track in reversedArray) {
            if(row > [[tracksController arrangedObjects] count])
                row--;
            
            [track setPlaylistIndex:[NSNumber numberWithInteger:row]];
        }
        
        
        [tracksController rearrangeObjects];
        [tracksTableView reloadData];
        
        // submit changes to server, this uses createPlaylist behind the scenes since we can reorder with it
        if (playlist.server) {
            [playlist.server updatePlaylistWithID: playlist.itemId tracks: tracksController.arrangedObjects];
        }
    }
    
    return YES;
}




#pragma mark -
#pragma mark Tracks NSTableView DataSource (Rating)

- (void)tableView:(NSTableView *)aTableView setObjectValue:(id)anObject forTableColumn:(NSTableColumn *)aTableColumn row:(NSInteger)rowIndex {
    if(aTableView == tracksTableView) {
        if([[aTableColumn identifier] isEqualToString:@"rating"]) {
            
            NSInteger selectedRow = [tracksTableView selectedRow];
            if(selectedRow != -1) {
                SBTrack *clickedTrack = [[tracksController arrangedObjects] objectAtIndex:selectedRow];
                
                if(clickedTrack) {
                    
                    NSInteger rating = [anObject intValue];
                    NSString *trackID = [clickedTrack itemId];
                    
                    [clickedTrack.server setRating:rating forID:trackID];
                }
            }
        }
    }
}



#pragma mark -
#pragma mark UI Validator

- (BOOL)validateUserInterfaceItem: (id<NSValidatedUserInterfaceItem>) item {
    SEL action = [item action];
    
    NSInteger tracksSelected = tracksTableView.selectedRowIndexes.count;
    
    SBSelectedRowStatus selectedTrackRowStatus = 0;
    selectedTrackRowStatus = [self selectedRowStatus: tracksController.arrangedObjects selectedIndices: tracksTableView.selectedRowIndexes];
    
    if (action == @selector(addSelectedToTracklist:)
        || action == @selector(playSelected:)
        || action == @selector(createNewLocalPlaylistWithSelectedTracks:)) {
        return tracksSelected;
    }
    
    if (action == @selector(showSelectedInFinder:)) {
        return selectedTrackRowStatus & SBSelectedRowShowableInFinder;
    }
    
    if (action == @selector(downloadSelected:)) {
        return selectedTrackRowStatus & SBSelectedRowDownloadable;
    }

    return YES;
}


@end

File: Submariner/SBPlaylistController.swift
//
//  SBPlaylistController.swift
//  Submariner
//
//  Created by Calvin Buckley on 2024-06-03.
//
//  Copyright (c) 2024 Calvin Buckley
//  SPDX-License-Identifier: BSD-3-Clause
//  

import Cocoa

@objc(SBPlaylistController) class SBPlaylistController: SBViewController, NSTableViewDelegate, NSTableViewDataSource, NSUserInterfaceValidations {
    // nulled means this playlist got deleted and hopefully the UI switched away from this VC correctly
    @objc var playlist: SBPlaylist! {
        didSet {
            if let playlistName = playlist?.resourceName {
                self.title = "Playlist \"\(playlistName)\""
            } else {
                self.title = "No Playlist"
            }
        }
    }
    @objc var playlistSortDescriptors: [NSSortDescriptor] = []
    @objc var databaseController: SBDatabaseController!
    
    @IBOutlet var tracksTableView: SBTableView!
    @IBOutlet var tracksController: NSArrayController!
    
    override class func nibName() -> String! {
        "Playlist"
    }
    
    private var selectionObserver: NSKeyValueObservation?
    
    override func loadView() {
        super.loadView()
        
        tracksTableView.registerForDraggedTypes([.libraryItem, .libraryItems])
        
        selectionObserver = tracksController.observe(\.selectedObjects) { ac, value in
            if self.view.window != nil {
                NotificationCenter.default.post(name: .SBTrackSelectionChanged, object: self.tracksController.selectedObjects)
            }
        }
    }
    
    override func viewDidAppear() {
        NotificationCenter.default.post(name: .SBTrackSelectionChanged, object: self.tracksController.selectedObjects)
    }
    
    // #MARK: - Typed Wrappers
    
    var tracks: [SBTrack]! {
        tracksController.arrangedObjects as? [SBTrack]
    }
    
    override var selectedTracks: [SBTrack]! {
        tracksController.selectedObjects as? [SBTrack]
    }
    
    // #MARK: - Actions
    
    @IBAction func trackDoubleClick(_ sender: Any) {
        if tracksTableView.selectedRow != -1 {
            SBPlayer.sharedInstance().play(tracks: tracks, startingAt: tracksTableView.selectedRow)
        }
    }
    
    @IBAction func playSelected(_ sender: Any) {
        self.trackDoubleClick(sender)
    }
    
    @IBAction func addSelectedToTracklist(_ sender: Any) {
        SBPlayer.sharedInstance().add(tracks: selectedTracks, replace: false)
    }
    
    @IBAction func removeTrack(_ sender: Any) {
        if selectedTracks.count > 0 {
            let alert = NSAlert()
            let removeButton = alert.addButton(withTitle: "Remove")
            removeButton.hasDestructiveAction = true
            alert.addButton(withTitle: "Cancel")
            alert.messageText = "Remove the selected tracks?"
            alert.informativeText = "The selected tracks will be removed from this playlist."
            alert.alertStyle = .warning
            
            alert.beginSheetModal(for: self.view.window!) { response in
                if response != .alertFirstButtonReturn {
                    return
                }
                
                let indices = self.tracksController.selectionIndexes
                self.playlist.remove(indices: indices)
                
                // update on server all at once
                guard let server = self.playlist.server, let playlistID = self.playlist.itemId else {
                    return
                }
                let indexArray = Array(indices)
                server.updatePlaylist(ID: playlistID, removing: indexArray)
            }
        }
    }
    
    @IBAction func delete(_ sender: Any) {
        self.removeTrack(sender)
    }
    
    @IBAction func showSelectedInFinder(_ sender: Any) {
        if tracksTableView.selectedRow != -1 {
            self.showTracksInFinder(tracks, selectedIndices: tracksTableView.selectedRowIndexes)
        }
    }
    
    @IBAction func downloadSelected(_ sender: Any) {
        if tracksTableView.selectedRow != -1 {
            self.downloadTracks(tracks, selectedIndices: tracksTableView.selectedRowIndexes, databaseController: databaseController)
        }
    }
    
    @IBAction func createNewLocalPlaylistWithSelectedTracks(_ sender: Any) {
        if tracksTableView.selectedRow != -1 {
            self.createLocalPlaylist(withSelected: tracks, selectedIndices: tracksTableView.selectedRowIndexes, databaseController: databaseController)
        }
    }
    
    @IBAction func showSelectedInLibrary(_ sender: Any) {
        if let first = selectedTracks.first {
            databaseController.go(to: first)
        }
    }
    
    // #MARK: - NSTableView (Drag & Drop)
    
    func tableView(_ tableView: NSTableView, pasteboardWriterForRow row: Int) -> (any NSPasteboardWriting)? {
        if tableView == tracksTableView {
            let track = tracks[row]
            return SBLibraryItemPasteboardWriter(item: track, index: row)
        }
        return nil
    }
    
    func tableView(_ tableView: NSTableView, validateDrop info: any NSDraggingInfo, proposedRow row: Int, proposedDropOperation dropOperation: NSTableView.DropOperation) -> NSDragOperation {
        guard row != -1 && dropOperation == .above else {
            return []
        }
        
        if let sourceTable = info.draggingSource as? SBTableView, sourceTable == tracksTableView {
            return .move
        } else if info.draggingPasteboard.libraryItems() != nil {
            return .copy
        }
        return []
    }
    
    func tableView(_ tableView: NSTableView, acceptDrop info: any NSDraggingInfo, row: Int, dropOperation: NSTableView.DropOperation) -> Bool {
        // XXX: For some reason, draggingSourceOperationMask has all bits set?
        if let sourceTable = info.draggingSource as? SBTableView, sourceTable == tracksTableView {
            let indices = info.draggingPasteboard.rowIndices()
            if let newIndexSet = playlist.moveTracks(fromOffsets: indices, toOffset: row) {
                tracksTableView.selectRowIndexes(newIndexSet, byExtendingSelection: false)
            }
        } else if let tracks = info.draggingPasteboard.libraryItems(managedObjectContext: self.managedObjectContext) {
            playlist.add(tracks: tracks, at: row)
        }
        
        tracksController.rearrangeObjects()
        tracksTableView.reloadData()
        
        // submit changes to server, this uses createPlaylist behind the scenes since we can reorder with it
        if let server = playlist.server, let playlistID = playlist.itemId {
            server.updatePlaylist(ID: playlistID, tracks: tracks)
        }
        return true
    }
    
    // #MARK: - NSTableView (Rating)
    
    func tableView(_ tableView: NSTableView, setObjectValue object: Any?, for tableColumn: NSTableColumn?, row: Int) {
        // previous version checked selectedRow, which seems redundant if we have row passed
        if tableView == tracksTableView && tableColumn?.identifier.rawValue == "rating" {
            let clickedTrack = tracks[row]
            
            guard let ratingNumber = object as? NSNumber else {
                return
            }
            
            clickedTrack.server?.setRating(ratingNumber.intValue, id: clickedTrack.itemId!)
        }
    }
    
    // #MARK: - UI Validator
    
    func validateUserInterfaceItem(_ item: any NSValidatedUserInterfaceItem) -> Bool {
        let tracksSelected = selectedTracks.count
        let selectedTrackRowStatus = self.selectedRowStatus(tracks, selectedIndices: tracksController.selectionIndexes)
        
        switch item.action {
        case #selector(SBPlaylistController.addSelectedToTracklist(_:)),
            #selector(SBPlaylistController.playSelected(_:)),
            #selector(SBPlaylistController.trackDoubleClick(_:)),
            #selector(SBPlaylistController.removeTrack(_:)),
            #selector(SBPlaylistController.delete(_:)),
            #selector(SBPlaylistController.createNewLocalPlaylistWithSelectedTracks(_:)):
            return tracksSelected > 0
        case #selector(SBPlaylistController.showSelectedInFinder(_:)):
            return selectedTrackRowStatus.contains(.showableInFinder)
        case #selector(SBPlaylistController.downloadSelected(_:)):
            return selectedTrackRowStatus.contains(.downloadable)
        case #selector(SBPlaylistController.showSelectedInLibrary(_:)):
            return tracksSelected == 1
        default:
            return true
        }
    }
}
